crates/engine/local/src/service.rs

```diff
@@ -0,0 +1,322 @@
+    }
+}
+
+/// The internal state of the [`LocalEngineService`]'s polling task
+#[derive(Default)]
+struct RequestState {
+    ready: bool,
+    task: Option<BoxFuture<'static, eyre::Result<B256>>>,
+}
+
+impl<N> LocalEngineService<N>
+where
+    N: EngineNodeTypes,
+{
+    /// Constructor for [`LocalEngineService`].
+    pub fn new(
+        payload_builder: PayloadBuilderHandle<N::Engine>,
+        payload_attributes: <N::Engine as PayloadTypes>::PayloadAttributes,
+        provider: ProviderFactory<N>,
+        pruner: Pruner<N::DB, ProviderFactory<N>>,
+        head: B256,
+        mode: MiningMode,
+    ) -> Self {
+        let persistence_handle = PersistenceHandle::spawn_service(provider, pruner);
+
+        Self {
+            payload_builder,
+            payload_attributes,
+            persistence_handle,
+            head,
+            mode,
+            payload_request_state: RequestState::default(),
+        }
+    }
+
+    /// Spawn a [`LocalEngineService`] on a tokio green thread.
+    pub fn spawn_new(
+        payload_builder: PayloadBuilderHandle<N::Engine>,
+        payload_attributes: <N::Engine as PayloadTypes>::PayloadAttributes,
+        provider: ProviderFactory<N>,
+        pruner: Pruner<N::DB, ProviderFactory<N>>,
+        head: B256,
+        mode: MiningMode,
+    ) {
+        let engine = Self::new(payload_builder, payload_attributes, provider, pruner, head, mode);
+
+        // Spawn the engine
+        tokio::spawn(engine);
+    }
+}
+
+/// Run the [`LocalEnginService`] as a Future. The service will poll the payload builder
+/// with two varying modes, [`MiningMode::Instant`] or [`MiningMode::Interval`]
+/// which will respectively either execute the block as soon as it finds a
+/// transaction in the pool or build the block based on an interval.
+impl<N> Future for LocalEngineService<N>
+where
+    N: EngineNodeTypes,
+{
+    type Output = ();
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.get_mut();
+        const MAX_RETRIES: usize = 5;
+        let mut retries = 0;
+        loop {
+            // Wait for the interval or the pool to receive a transaction
+            if !this.payload_request_state.ready {
+                ready!(pin!(&mut this.mode).poll(cx));
+                this.payload_request_state.ready = true;
+            }
+
+            if this.payload_request_state.task.is_none() {
+                let head = this.head;
+                let attributes = this.payload_attributes.clone();
+                let payload_builder = this.payload_builder.clone();
+                let persistence_handle = this.persistence_handle.clone();
+
+                let task = Box::pin(async move {
+                    let attr = <N::Engine as PayloadTypes>::PayloadBuilderAttributes::try_new(
+                        head, attributes,
+                    )
+                    .map_err(|_| eyre::eyre!("failed to fetch payload attributes"))?;
```

Comment:
I can't figure out how to update the attributes here. My current understanding was:
- Pass an initial `PayloadAttributes ` to the engine
- Use this to create the `PayloadBuilderAttributes`
- Update the `PayloadAttributes` once block building is done

However I can't figure out how to update these attributes. Maybe I need to add a setter for them?